feat: add dynamodb table and provider for tag and path revalidation

kevin-mitchell opened the pull request:
This diff adds the DynamoDB table and supporting data provider function to support Next.js path and tag revalition. This feature is supported by `open-next` (read more about it here: https://open-next.js.org/inner_workings/isr#tags) and is inspired by https://github.com/sst/sst/compare/v2.27.0...v2.28.0

The changes include

* Adding a new DynamoDB table which stores time stamps, tags, and paths for revalidation consideration
* Adds a function (`dynamodb-provider`) which populates the DynamoDB table on deployment with the initial set of tags / paths / timestamps generated in the Next.js build phase
* **Bumps `aws-cdk-lib` to 2.95.1**, matching the SST version for access to `DynamoDB.TableV2`
* Adds a simple description to the "primary" server handler

A bit more info:

### Bump of `aws-cdk-lib` 2.95.1

I did this because SST uses the DynamoDB `TableV2` construct and I thought it would be best to match.

### Description of ServerFunction

I added a description because at least for me the function names end up mangled (I actually think it’d be great to clean these up, but I frankly don’t know the best practice here from IaC / AWS perspective) and it’s nice to be able to easily identify different functions, in particular the server handler, in the AWS console for debugging.

### Non breaking change, but…

From what I can tell this is a non-breaking change, however this brings me back to a general question about how we should be managing versions and default open-next build commands… SST uses `2.2.1` as the default version of `open-next` in the build command.

#143

kevin-mitchell commented:
I'm marking this as a draft as I need to figure out if it's actually fully working (e.g. the expected functionality is working), and also importantly make sure it's safe to land this if somebody is using an older version of open-next (e.g. whatever the current default open-next build command is).

It is at least "working" in that the table is created on deployment, and data is making it's way into the table (see screenshot)
<img width="1046" alt="Screenshot 2023-10-11 at 5 47 37 PM" src="https://github.com/jetbridge/cdk-nextjs/assets/21270844/52a8d605-5ac6-413c-8d58-1a0614d03a24">

I'm also not sure if there are issues / concerns with me bumping the aws cdk version. From what I can tell it's not creating any issues, but that's based on my fairly unscientific testing. 

FYI @bestickley :)

bestickley commented on src/NextjsRevalidation.ts:
```diff
@@ -94,13 +108,94 @@ export class NextjsRevalidation extends Construct {
+    return new Table(this, 'RevalidationTable', {
+      partitionKey: { name: 'tag', type: AttributeType.STRING },
+      sortKey: { name: 'path', type: AttributeType.STRING },
+      pointInTimeRecovery: true,
```
I think point in time recovery should be disabled by default to save cost. This is a cache, so cost of losing data is low. PITR costs $0.20/GB-month which isn't much. I could be persuaded otherwise though.

kevin-mitchell replied:
I actually agree 100% - I tried to see if this was somehow included in free tier, but I don't think so. This is how SST has the table configured, but I guess they are less shy about the cost. $0.20 isn't much, but it's nice to not have to think about it and like you said, this is cache. 

I think if the data in this table is completely missing or corupted `revalidateTag` and `revalidatePath` will not work for SSG pages (based on  https://discord.com/channels/983865673656705025/1027265626085019769/1161575754727968801 ), BUT the site would still work and the table could be repopulated / redeployed easily it seems.

bestickley commented on src/NextjsServer.ts:
```diff
@@ -117,6 +117,7 @@ export class NextjsServer extends Construct {
       ...getCommonFunctionProps(this),
       code: Code.fromBucket(asset.bucket, asset.s3ObjectKey),
       handler: 'index.handler',
+      description: 'Next.js server handler',
```
Great idea!

bestickley replied:
Should we do this for all functions?

kevin-mitchell replied:
I think most of the functions we're explicitly creating have descriptions now. Some of the ancillary functions do not, but to me that seems "fine" given most of the time when I'm hunting for a function to debug it's either server handler (99% of the time), the queue handler, or the image handler, and those are now all easily identifiable from the description. 

<img width="1148" alt="Screenshot 2023-10-12 at 9 27 41 AM" src="https://github.com/jetbridge/cdk-nextjs/assets/21270844/003d3aa6-9a39-41a1-a55c-df6b2f8e1324">

bestickley commented on src/NextjsRevalidation.ts:
```diff
@@ -78,7 +92,7 @@ export class NextjsRevalidation extends Construct {
   private createFunction(): NodejsFunction {
     const nodejsFnProps = getCommonNodejsFunctionProps(this);
-    const fn = new NodejsFunction(this, 'Fn', {
+    const fn = new NodejsFunction(this, 'QueueRevalidationFn', {
```
I made the id of this only `Fn` so that is deduplicates the term `Revalidation` in the name of the resource since the parent construct's name is `Revalidation`, it gets added twice. So I would prefer this to be `QueueFn` so that the ultimate resource name would be `RevalidationQueueFn` instead of `RevalidationQueueRevalidationFn`.

bestickley replied:
With `Fn`, this is an example of a current resource name: `bui1-stickb-ui-NextjsRevalidationFn329BA096-CKuulcO3k6bB`

kevin-mitchell replied:
Makes sense for sure. I'll change to `QueueFn` and `DynamoDBProviderFn`.

RE: names in general, for other projects I work on (unrelated to Next.js or this repo), I have function names like

```
prod-public-gateway-authorizer
prod-s3-publisher-handler 
prod-introspection-service
```

I normally prefix everything with a "stage" (even though I run different stages in different accounts, I still really like to be able to quickly / visually see "prod" or "dev", etc), then the names are very flat. I frankly don't know if this is "good" though, as I mentioned I'm not sure on best practice here, it just _looks_ tidy :).

For now I'm going to just strip "Revalidation" out, end up with (in my case) things like `Cdk2Stack-WebRevalidationQueueFn37B09F4F-zUasGdL53ANQ`

bestickley commented on .projenrc.ts:
```diff
@@ -39,7 +39,7 @@ const project = new awscdk.AwsCdkConstructLibrary({
   // dependency config
   jsiiVersion: '~5.0.0',
-  cdkVersion: '2.93.0',
+  cdkVersion: '2.95.1',
```
I have no problem with bumping CDK version. I think we bump to latest if you'd like.

kevin-mitchell replied:
Bumped to the latest version, seems fine as far as I can tell!

bestickley reviewed (changes requested):
Great job, @kevin-mitchell! Let's get this validated to make sure it works.

kevin-mitchell commented:
So this is working in the following ways:

1. The initial "dynamodb-provider" is writing to the DynamoDB table on deployment - to me this is a sign that the basic wiring of the table is correct and that communication with the table is possible
2. Making a GET request to `/api/revalidate` from the [https://github.com/vercel/app-playground/blob/main/app/api/revalidate/route.ts](https://github.com/vercel/app-playground/blob/main/app/api/revalidate/route.ts) Vercel App Playground will trigger writes to the DynamoDB table, a good indication I believe that the wiring between the server handler and the DynamoDB table is good / working

THAT SAID, I'm not actually sure that the revalidation feature is working, but I'm not sure it's not user error on my part.

I added a simple page in `/app/isr/cdkstandalone/600sec` that looks like this:

```
export const revalidate = 600;

export default function Page() {
  return (
    <div className="prose prose-sm prose-invert max-w-none">
      <h1 className="text-xl font-bold">600 second ISR config</h1>
      <strong>Rendered at {new Date().toISOString()}</strong>
    </div>
  );
}
```

Then I deployed the app, and then made an API request to

`/api/revalidate?path=/isr/cdkstandalone/600sec` - I can see this is hitting the server in the logs, but as far as I can tell it's not actually triggering the route to be revalidated. This could be user error on my part.

To put this another way, this looks like it's working from the perspective of the DynamoDB table that I added, but I'm not sure it's actually _working_, or very possibly I'm misunderstanding something.

Regardless I think this is officially ready to review, but I would appreciate any input on the above or I'll spend more time looking into the actual open-next implementation to see what I might be missing or what I should be expecting. I need to check the actual DynamoDB query to see if something is wrong there, I noticed in the DDB table there is a prefix (a deployment hash or something?), perhaps something is going wrong there.

<img width="1050" alt="Screenshot 2023-10-13 at 4 46 45 PM" src="https://github.com/jetbridge/cdk-nextjs/assets/21270844/0c9be842-3a86-4a9f-971d-92209e92c975">

bestickley commented:
@kevin-mitchell, to answer your question I think you should see the DDB table update if the `revalidateAt` value is greater than the `LastModified` from S3.

From docs:
> When revalidateTag is called, we update the revalidatedAt value for each path and subpath associated with this tag.

> When we check if a page is stale, we check the revalidatedAt value for each record and the LastModified of this S3 cache objects . If revalidatedAt is greater than LastModified, we consider the page is stale.

Based on your test page above, I don't think that will work. Don't you need to use `fetch` with the special `next.tags` prop? See docs [here](https://nextjs.org/docs/app/building-your-application/caching#on-demand-revalidation). How I would test this is call `fetch("...", { next: { tags: ["my-tag"] } })` with some API that returns a date, and then keep refreshing page until the `export const revalidate = 600;` time has been reached and monitor DDB table throughout. I will try to get to it this week.

kevin-mitchell commented:
@bestickley I believe the "tag" in this case would be the URL - I don't _think_ `fetch` needs to be used. And by this I mean I'm 100% sure there are entries in the dynamodb table, the tag is just derived from the path.
 
<img width="993" alt="Screenshot 2023-10-19 at 5 36 09 PM" src="https://github.com/jetbridge/cdk-nextjs/assets/21270844/e0ad141d-6172-4b36-afd8-de551658ff32">

That said... I'm still not at all confident that I understand what the expected behavior is here. Even with this change, according to https://open-next.js.org/common_issues/isr#on-demand-revalidation (and this is semi-obvious I guess), the CloudFront cache needs to be cleared before you'd be able to see the results of calling `revalidateTag` or `revalidatePath` on the frontend.

If I manually invalidate the CloudFront cache, I'm seeing the static version of the page (in S3) regenerated regardless of whether or not the `revalidatePath/Tag` has been called or not. Leaving me to wonder what a good way of evaluating whether or not this is actually working is.

Also, I'll rebase this branch on top of your changes. I'm actually out this week so am a bit slow to respond, but hopefully with fresh eyes on Monday (or with any wisdom you might have!) this should be easy to clear up.

kevin-mitchell commented on src/NextjsBuild.ts:
```diff
@@ -101,7 +110,7 @@ export class NextjsBuild extends Construct {
   private build() {
     const buildPath = this.props.buildPath ?? this.props.nextjsPath;
-    const buildCommand = this.props.buildCommand ?? 'npx open-next@2 build';
+    const buildCommand = this.props.buildCommand ?? 'npx open-next@2.2 build';
```
@bestickley FYI I bumped this to 2.2 because some of the features of revalidation aren't available in 2.0/2.1 of `open-next`

bestickley replied:
@kevin-mitchell, thank you for this. I've noticed this too, that `npx open-next@2 build` doesn't use the latest version of 2.X for `open-next`. Do you know why this is? I thought specifying `@2` would do latest major version of 2, right?

kevin-mitchell replied:
I believe (though it's not 100% clear to me if this is the correct behavior) `@2` will use any version of 2 in `npx` cache (`~/.npm/_npx/` on Mac). I haven't spent a ton of time testing so I'm not sure if being more explicit (e.g. `@^2` or perhaps `@~2.2`) would do the trick, or if this just works for me just now because it wasn't in the NPX cache and https://github.com/npm/cli/issues/6804 is a legit issue.

bestickley replied:
@kevin-mitchell, could you change this to `npx open-next@^2 build` so we don't have to keep updating this? Make sure to update comment to.

kevin-mitchell replied:
yep, done!

bestickley commented:
Regardless of what "tag" is in DDB table, have you read the Next.js docs on [revalidateTag](https://nextjs.org/docs/app/api-reference/functions/revalidateTag)? Look in the parameters section and see how it says:

> tag: A string representing the cache tag associated with the data you want to revalidate. Must be less than or equal to 256 characters.

And demo usage is using `fetch`. So in order to test, we need to use that function. If you haven't, please read the [Next.js Caching Docs](https://nextjs.org/docs/app/building-your-application/caching). The caching we're working on is specifically the [Data Cache](https://nextjs.org/docs/app/building-your-application/caching#data-cache).

In your test page (/app/isr/cdkstandalone/600sec), that is only showcasing [Time-based revalidation](https://nextjs.org/docs/app/building-your-application/caching#time-based-revalidation) I think. This feature is about [On-demand Revalidation](https://nextjs.org/docs/app/building-your-application/caching#on-demand-revalidation).

With that said, I'm confused that there aren't any tags in your DDB table. However, you're not creating any since you're not using `fetch` with a `tag` parameters. There is also the `revalidatePath` function so maybe in that case open-next's ddb revalidate function just uses the `tag` partition key to add the path. (If I were them I would have used generic pk for partition key and sk for sort key so it's less confusing.)

So could you try testing with that context? Please let me know if anything doesn't make sense. TBH, I don't have much experience with `revalidateTag` or `revalidatePath` but this is based off my research.

kevin-mitchell commented:
@bestickley 

> With that said, I'm confused that there aren't any tags in your DDB table. However, you're not creating any since you're not using `fetch` with a `tag` parameters...

Sorry I might have been unclear here specifically - there are many tags and paths in DDB:

<img width="1015" alt="Screenshot 2023-10-21 at 10 22 59 PM" src="https://github.com/jetbridge/cdk-nextjs/assets/21270844/abe8010f-442e-4c25-85c9-034f10be57bf">

These seem to be generated both by page views as well as actually directly using `fetch`. I had (wrongly?) assumed that the  `open-next` cache handler ([specifically around this area of the code](https://github.com/serverless-stack/open-next/blob/24c5166377346bed1a029fb2223cf7f3b52a10d3/packages/open-next/src/adapters/cache.ts#L138)) would work the same for either pages or fetch data, in the end I figure the logic was "check in S3 for object, if DDB revalidatedAt tag / path matches condition then refetch the page, else server the version from S3". I'm still not sure that's not the case frankly, BUT I went ahead and used `fetch` as it is certainly most clear given the `tag` is directly supplied.

From what I can tell this is working well.

[I recorded a short "sanity check" for myself talking through the steps I took to actually test / confirm things are working, video is here if you care to listen to me ramble :).](https://drive.google.com/file/d/1l30_67Hqlg0kfH4V0n3ndEg_qUZE3r3v/view?usp=share_link)

I will (Monday) play around with this a bit more and with fresh eyes actually step through the open-next code to see how the execution path is different for `fetch` vs page cache (because to reiterate, without using `fetch`, normal page views still generated tags, e.g. `<deploymentkey>/whatever/page/path...` as a `tag` in DDB.

Still overall I think this is probably ready to go from a functionality perspective.

kevin-mitchell commented:
OK I spent a bit more time today looking at this. One thing I'm fairly certain of is that the revalidate functionality works perfectly fine for ISR page cache, not just `fetch`. I tested this and confirmed it works with the app router. I can make a simple page, without using `fetch`, export `revalidation = 600`, confirm that the page is cached and step through the debug logs to confirm that the version of the page in `_cache` is being served when I invalidate the CloudFront cache, and then confirm I can change this behavior by manually calling the `revalidatePath` function with the matching path. 

---

One thing I'm less certain of and am questioning (though to be clear I don't think the scope of this PR should get expanded here, just something I noticed), is using the App router specifically how static site generation is intended to work, or if it's frankly even a feature of Next.js (see [this post on reddit](https://www.reddit.com/r/nextjs/comments/144lsbj/app_router_can_you_do_static_site_generation/). 

When we deploy with `cdk-nextjs`, the build cache files end up in `<s3 bucket>/cache` ([controlled by `NEXTJS_CACHE_DIR`](https://github.com/jetbridge/cdk-nextjs/blob/main/src/NextjsStaticAssets.ts#L77)). But I'm not seeing those files ever being served. So if for example I have this page:

```
// <nextjs root>/app/1norevalidatekam/page.tsx
export default function Page() {
  return (
    <div className="prose prose-sm prose-invert max-w-none">
      <h1 className="text-xl font-bold">NO REVALIDATE</h1>
      <strong>Rendered at {new Date().toISOString()}</strong>
    </div>
  );
}
```

I would expect (**correct?**) that the page generated at build time would be served. At least that's what I'd think would happen.

Instead, it seems like the cache handler checks for the page in `_cache` (note the underscore), defined in our constant `CACHE_BUCKET_KEY_PREFIX`.

I've read (and re-read) https://open-next.js.org/advanced/options#reusing-same-bucket-for-asset-and-cache and to me it seems like the `_cache` and `cache` directory should be the same thing? That way the original cache files are build time are used by the cache system, vs (as far as I can tell) not used at all?

(it's possible I'm missing something with a CloudFront original or something that is making a distinction between these in a way that would require they be separate folders in S3)

bestickley commented:
Thanks for creating a separate issue and PR for caching issue. Wonderful video. Clearly demonstrates this feature working. I'll merge once you resolve last comment.

bestickley commented on src/NextjsRevalidation.ts:
```diff
@@ -27,25 +31,34 @@ export interface NextjsRevalidationProps extends NextjsBaseProps {
+    this.tableFunction = this.createRevalidationInsertFunction(this.table);
+
+    this.props.serverFunction.lambdaFunction.addEnvironment('CACHE_DYNAMO_TABLE', this.table.tableName);
+    this.table.grantReadWriteData(this.props.serverFunction.lambdaFunction.role!);
```
If for whatever reason the role is `undefined`, I don't want this to error. Could you change to:
```ts
if (this.props.serverFunction.lambdaFunction.role) {
  this.table.grantReadWriteData(this.props.serverFunction.lambdaFunction.role);
}
```

bestickley commented on src/NextjsRevalidation.ts:
```diff
@@ -74,16 +87,83 @@ export class NextjsRevalidation extends Construct {
 
   private createFunction(): LambdaFunction {
```
`createQueueFunction` now?

bestickley commented on src/NextjsRevalidation.ts:
```diff
@@ -74,16 +87,83 @@ export class NextjsRevalidation extends Construct {
+        handler: 'index.handler',
+        description: 'Next.js Revalidation DynamoDB Provider',
+        timeout: Duration.minutes(15),
+        initialPolicy: [
```
Instead of specifying an `initialPolicy`, what do you think about adding:
```
revalidationTable.grantReadWrite(insertFn);
```
below? Keeps it simpler IMO.

bestickley commented on src/NextjsRevalidation.ts:
```diff
@@ -74,16 +87,83 @@ export class NextjsRevalidation extends Construct {
+        code: Code.fromAsset(this.props.nextBuild.nextRevalidateDynamoDBProviderFnDir),
+        handler: 'index.handler',
+        description: 'Next.js Revalidation DynamoDB Provider',
+        timeout: Duration.minutes(15),
```
Could we decrease timeout here? to say 1 min? Doubt we need 15

kevin-mitchell replied:
Done, but only comment here is I based this off of the SST construct and they went with 15 minutes here. They do this in a few different places, so maybe this is just their preferred default, but in this case I really don't have a great idea of how long a reasonable timeout is. I agree 1 minute seems reasonable, but also there are quite a few factors as I understand it that can impact startup time, e.g. Lambda size... 1 minute still seems reasonable, but maybe just something to keep an eye on?

bestickley replied:
yes, let's keep an eye on it. good call.
